src/coreclr/System.Private.CoreLib/src/System/Array.CoreCLR.cs

```diff
@@ -95,8 +96,311 @@ private static unsafe void CopyImpl(Array sourceArray, int sourceIndex, Array de
+        {
+            Debug.Assert(sourceArray.Rank == destinationArray.Rank);
+
+            void* srcTH = RuntimeHelpers.GetMethodTable(sourceArray)->ElementType;
+            void* destTH = RuntimeHelpers.GetMethodTable(destinationArray)->ElementType;
+            AssignArrayEnum r = CanAssignArrayType(srcTH, destTH);
+
+            if (r == AssignArrayEnum.AssignWrongType)
+                ThrowHelper.ThrowArrayTypeMismatchException_CantAssignType();
+
+            if (length > 0)
+            {
+                switch (r)
+                {
+                    case AssignArrayEnum.AssignUnboxValueClass:
+                        CopyImplUnBoxEachElement(sourceArray, sourceIndex, destinationArray, destinationIndex, length);
+                        break;
+
+                    case AssignArrayEnum.AssignBoxValueClassOrPrimitive:
+                        CopyImplBoxEachElement(sourceArray, sourceIndex, destinationArray, destinationIndex, length);
+                        break;
+
+                    case AssignArrayEnum.AssignMustCast:
+                        CopyImplCastCheckEachElement(sourceArray, sourceIndex, destinationArray, destinationIndex, length);
+                        break;
+
+                    case AssignArrayEnum.AssignPrimitiveWiden:
+                        CopyImplPrimitiveWiden(sourceArray, sourceIndex, destinationArray, destinationIndex, length);
+                        break;
+
+                    default:
+                        Debug.Fail("Fell through switch in Array.Copy!");
+                        break;
+                }
+            }
+        }
+
+        private enum AssignArrayEnum
+        {
+            AssignWrongType,
+            AssignMustCast,
+            AssignBoxValueClassOrPrimitive,
+            AssignUnboxValueClass,
+            AssignPrimitiveWiden,
+        }
+
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Array_CanAssignArrayType")]
+        private static unsafe partial AssignArrayEnum CanAssignArrayType(void* srcTH, void* dstTH);
+
+        // Unboxes from an Object[] into a value class or primitive array.
+        private static unsafe void CopyImplUnBoxEachElement(Array sourceArray, int sourceIndex, Array destinationArray, int destinationIndex, int length)
+        {
+            MethodTable* pDestArrayMT = RuntimeHelpers.GetMethodTable(destinationArray);
+            TypeHandle destTH = pDestArrayMT->GetArrayElementTypeHandle();
+
+            Debug.Assert(!destTH.IsTypeDesc && destTH.AsMethodTable()->IsValueType);
+            Debug.Assert(!RuntimeHelpers.GetMethodTable(sourceArray)->GetArrayElementTypeHandle().AsMethodTable()->IsValueType);
+
+            MethodTable* pDestMT = destTH.AsMethodTable();
+            nuint destSize = pDestArrayMT->ComponentSize;
+            ref object? srcData = ref Unsafe.Add(ref MemoryMarshal.GetArrayDataReference(Unsafe.As<object?[]>(sourceArray)), sourceIndex);
```

Comment:
What guarantees that the arrays are single-dimensional arrays? IIRC, Array.Copy works on  multidimensional arrays too.